Molecules/Taskの作成

caramelribon opened the pull request:
## Molucules/Taskの作成1st

- [x] お手本通りの見た目であるか　ホバー時の挙動、レスポンシブ対応など要確認
- [x] ブランチ名、ファイル名、クラス名、インデントなど、コーディング規則を逸脱していないか
- [x] 記述の重複はないか

caramelribon committed:
Molecules/Task

first commit

Higashiji commented on src/components/Molecules/Task/index.jsx:
```diff
@@ -0,0 +1,74 @@
+import TEXT from "../../../variables/texts";
+
+const Task = (props) => {
+  // inputを表示するか、タスク名を表示するかの状態管理
+  const [isVisible, setVisibility] = useState(false);
+  // タスクが編集中かどうかの状態管理
+  const [isEdit, setEdit] = useState(props.defalutEdit);
```
タスクが編集中 ＝ inputを表示する
なので、２つのstateを isEdit の方にまとめてもいいと思います！

Higashiji replied:
命名に関してですが、
state名が `isEdit` であればsetterの方も `setIsEdit` という命名にした方が対応関係がわかりやすくていいと思います。

また、この場合 isEditing としたほうが「現在編集中である」ということが伝わりやすいかもしれません！

Higashiji replied:
その場合propsも `defaultIsEditing` とかのほうが良さそうです。

Higashiji commented on src/components/Molecules/Task/index.jsx:
```diff
@@ -0,0 +1,74 @@
+import COLOR from "../../../variables/color";
+import TEXT from "../../../variables/texts";
+
+const Task = (props) => {
```
```suggestion
const Task = ({
  defaultValue,
  defaultEdit
  // その他にもいろいろ
}) => {
```
分割代入を使うと、どのようなpropsが存在するのかがわかりやすくなり見通しが良くなると思います。
参考： https://dezanari.com/react-component-props-object/

Higashiji reviewed (commented):
見た目は完璧だと思います！

- チェックマークをつけたときのイベントハンドラ
- 編集を完了したときのイベントハンドラ

上記二点の関数は親要素からpropsで渡すようにしてください。
編集完了時のイベントハンドラに関しては、入力された文字列を引数として渡すようにしてください。
今作っているパーツがこの先どのように使われるかを見に行くと、どのような実装が必要なのかをイメージしやすいかと思います！

caramelribon committed:
propsで渡すものと変数の名前の修正

親要素からpropsで渡すものをキレイにまとめました

useStateのstate名とsetter名をisEding setInEditingに、defalutEditをdefaultIsEditingに修正しました

caramelribon commented:
## Molucules/Taskの作成2nd

指摘して頂いたところを修正しました。

- [x] お手本通りの見た目であるか　ホバー時の挙動、レスポンシブ対応など要確認
- [x] ブランチ名、ファイル名、クラス名、インデントなど、コーディング規則を逸脱していないか
- [x] 記述の重複はないか

Higashiji commented on src/components/Molecules/Task/index.jsx:
```diff
@@ -0,0 +1,73 @@
+  taskName,
+}) => {
+  // タスクが編集中かどうかの状態管理
+  const [isEditing, setIsEditing] = useState({ defalutIsEditing });
```
```suggestion
  const [isEditing, setIsEditing] = useState( defalutIsEditing );
```

この代入の仕方だと下の画像のような感じになって、Booleanの値が渡せないので修正をお願いします！


![image](https://user-images.githubusercontent.com/83479175/189536628-42004f6d-8bbf-405c-96da-0a0c5f0196ea.png)

Higashiji commented on src/components/Molecules/Task/index.jsx:
```diff
@@ -0,0 +1,73 @@
+  // タスクが編集中かどうかの状態管理
+  const [isEditing, setIsEditing] = useState({ defalutIsEditing });
+
+  onTaskComplete = () => {
+    setIsEditing(false);
+    console.log("task completed");
+  };
```
Propsと同名の変数を定義すると、渡されたPropsを上書きすることになってしまうので修正が必要です。
Suggestionのように、このコンポーネント内で新たにハンドラを定義してその中でPropsとして渡された関数を実行する形にすれば、stateの変更とonTaskCompleteを両方実行できます。

要件にある `Task completed` を出力するのはこのコンポーネント内に書くのではなく、Storybook の Args において onTaskComplete に そのような関数を渡すことで再現できると思います。

```suggestion
 const handleCompleteTask = () => {
    setIsEditing(false);
    onTaskComplete();
  };
```

Higashiji commented on src/components/Molecules/Task/index.jsx:
```diff
@@ -0,0 +1,73 @@
+    setIsEditing(true);
+  };
+
+  onEditComplete = (taskName) => {
+    setIsEditing(false);
+    console.log(`taskname changed: ${taskName}`);
+    console.log({ taskName });
+  };
```
```suggestion
  const handleCompleteEdit = (taskName) => {
    setIsEditing(false);
    onEditComplete(TaskName);
  };
```
こちらもこんな感じに修正するのがいいかと！

Higashiji commented on src/components/Molecules/Task/index.jsx:
```diff
@@ -0,0 +1,73 @@
+import TEXT from "../../../variables/texts";
+
+const Task = ({
+  defalutValue,
+  defalutIsEditing,
+  onEditComplete,
+  onTaskComplete,
+  taskName,
```
defautlValue と TaskName は役割が被っていると思うので、どちらかでいいかと思います。

Higashiji commented on src/components/Molecules/Task/index.jsx:
```diff
@@ -0,0 +1,73 @@
+
+  return (
+    <StyledTask>
+      <CheckBox onClick={() => onTaskComplete()} />
```
```suggestion
      <CheckBox onClick={onTaskComplete} />
```
onTaskCompleteは引数を取らないのでこのようにかけます。

Higashiji commented on src/components/Molecules/Task/index.jsx:
```diff
@@ -0,0 +1,73 @@
+      ) : (
+        <StyledTaskNameArea>
+          <StyledTaskTitle>{defalutValue}</StyledTaskTitle>
+          <EditButton onClick={() => onEditStart()} />
```
```suggestion
          <EditButton onClick={onEditStart} />
```
こちらもこのようにしてもいいかと

caramelribon committed:
pref:propsの変数の見直し

役割が同じ変数を1つにまとめました

pref:propで渡す関数を上書きしないようにする

propsと同じ名前の変数を定義することによって、その関数が上書きされてしまうので新しい関数を作って、その中にpropsで渡される関数を入れることで上書きを防ぐ

caramelribon commented:
## Molucules/Taskの作成3rd

指摘して頂いたところを修正しました。

- [x] お手本通りの見た目であるか　ホバー時の挙動、レスポンシブ対応など要確認
- [x] ブランチ名、ファイル名、クラス名、インデントなど、コーディング規則を逸脱していないか
- [x] 記述の重複はないか

Higashiji reviewed (approved):
LGTMです！
